metadata-io/src/main/java/com/linkedin/metadata/graph/elastic/ESGraphQueryDAO.java

```diff
@@ -1469,4 +1366,704 @@ private Optional<BoolQueryBuilder> buildLineageGraphFiltersQuery(
+   * @param pageSize The number of documents to retrieve in each pagination request. Controls the
+   *     batch size for each Elasticsearch query.
+   * @param entitiesPerHopLimit Optional limit on the total number of entities to explore per hop.
+   * @return A stream of SearchResponse objects containing the lineage relationships. The stream
+   *     includes results from both incoming and outgoing edges, paginated using the search_after
+   *     mechanism.
+   * @throws ESQueryException If there is an error executing the Elasticsearch queries or if the
+   *     queries time out based on the configured timeout.
+   */
+  @WithSpan
+  private Stream<SearchResponse> executeGroupByLineageSearchQuery(
+      @Nonnull final OperationContext opContext,
+      @Nonnull final QueryBuilder query,
+      final LineageGraphFilters lineageGraphFilters,
+      final int pageSize,
+      boolean exploreMultiplePaths,
+      @Nullable final Integer entitiesPerHopLimit,
+      Set<Urn> originalEntityUrns) {
+
+    if (entitiesPerHopLimit != null && entitiesPerHopLimit == 0) {
+      log.warn("Requested 0 entities to explore per hop?");
+      return Stream.empty();
+    }
+
+    // Create source filter for outgoing relationships
+    BoolQueryBuilder sourceFilterQuery = QueryBuilders.boolQuery();
+    lineageGraphFilters
+        .streamEdgeInfo()
+        .filter(pair -> RelationshipDirection.OUTGOING.equals(pair.getValue().getDirection()))
+        .forEach(
+            pair ->
+                sourceFilterQuery.should(
+                    getAggregationFilter(pair, RelationshipDirection.OUTGOING)));
+
+    if (!sourceFilterQuery.should().isEmpty()) {
+      sourceFilterQuery.minimumShouldMatch(1);
+    }
+
+    // Create destination filter for incoming relationships
+    BoolQueryBuilder destFilterQuery = QueryBuilders.boolQuery();
+    lineageGraphFilters
+        .streamEdgeInfo()
+        .filter(pair -> RelationshipDirection.INCOMING.equals(pair.getValue().getDirection()))
+        .forEach(
+            pair ->
+                destFilterQuery.should(getAggregationFilter(pair, RelationshipDirection.INCOMING)));
+
+    if (!destFilterQuery.should().isEmpty()) {
+      destFilterQuery.minimumShouldMatch(1);
+    }
+
+    // Combine filters with the main query
+    BoolQueryBuilder outgoingQuery;
+    if (!sourceFilterQuery.should().isEmpty()) {
+      outgoingQuery = QueryBuilders.boolQuery();
+      outgoingQuery.filter(query);
+      outgoingQuery.filter(sourceFilterQuery);
+    } else {
+      outgoingQuery = null;
+    }
+
+    BoolQueryBuilder incomingQuery;
+    if (!destFilterQuery.should().isEmpty()) {
+      incomingQuery = QueryBuilders.boolQuery();
+      incomingQuery.filter(query);
+      incomingQuery.filter(destFilterQuery);
+    } else {
+      incomingQuery = null;
+    }
+
+    // Use CompletableFutures to run queries in parallel
+    List<CompletableFuture<List<SearchResponse>>> futures = new ArrayList<>();
```

Comment:
So basically splitting all of the topNHits aggs into parallel queries

Reply:
Yes, the two former aggs queries are executed in parallel.